src/social-network/defaults/GroupsValueRef.ts

```diff
@@ -1,12 +1,49 @@
-import { MessageCenter } from '../../utils/messages'
 import Services from '../../extension/service'
 import { SocialNetworkUI } from '../ui'
+import { ValueRef } from '@holoflows/kit/es'
+import { Group } from '../../database'
+import { GroupIdentifier, PreDefinedVirtualGroupNames, PersonIdentifier } from '../../database/type'
 
-export function InitGroupsValueRef(self: SocialNetworkUI, network: string) {
-    const ref = self.groupsRef
-    Services.People.queryUserGroups(network).then(p => (ref.value = p))
-    MessageCenter.on('newGroup', group => {
-        const old = ref.value.filter(x => !x.identifier.equals(group.identifier))
-        ref.value = [...old, group]
-    })
+// TODO:
+// groupIDs can be a part of network definitions
+export function InitGroupsValueRef(
+    self: SocialNetworkUI,
+    network: string,
+    groupIDs: string[] = [PreDefinedVirtualGroupNames.friends],
+) {
+    create(network, self.groupsRef, groupIDs)
+}
+
+async function query(network: string, ref: ValueRef<Group[]>) {
+    ref.value = await Services.People.queryUserGroups(network)
+}
+
+async function create(network: string, ref: ValueRef<Group[]>, groupIDs: string[]) {
+    type Pair = [PersonIdentifier, GroupIdentifier]
+    const [identities, groups] = await Promise.all([
+        Services.People.queryMyIdentities(network),
+        Services.People.queryUserGroups(network),
+    ])
+    const pairs = identities.reduce<Pair[]>(
```

Comment:
why not use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap?